File: Anul 2/Medii de Programare si Proiectare/mpp-proiect-java-Mothyque/backend/Server/src/main/java/org/example/repository/db/DBRepository.java
package org.example.repository.db;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.domain.Entity;
import org.example.domain.exception.PersistenceException;
import org.example.repository.Repository;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public abstract class DBRepository<ID, E extends Entity<ID>> implements Repository<ID, E>
{
    private final String url;
    private static final Logger logger = LogManager.getLogger();
    protected Connection connection;

    public DBRepository(String url)
    {
        this.url = url;
        openConnection();
        logger.info("Initializing DBRepository with URL: " + url);
    }

    protected Connection getConnection() throws SQLException
    {
        if(connection == null || connection.isClosed())
        {
            logger.debug("Establishing new database connection");
            connection = DriverManager.getConnection(url);
        }
        return connection;
    }

    @Override
    public Iterable<E> findAll()
    {
        logger.traceEntry("Finding all entities");
        List<E> entities = new ArrayList<>();
        String sql = getFindAllSQL();
        try (PreparedStatement statement = getConnection().prepareStatement(sql);
             ResultSet rs = statement.executeQuery())
        {
            while (rs.next())
            {
                E entity = extractEntity(rs);
                entities.add(entity);
            }
        }
        catch (SQLException e)
        {
            logger.error("Error finding all entities", e);
            throw new PersistenceException("Error finding all entities", e);
        }
        logger.traceExit("Found " + entities.size() + " entities");
        return entities;
    }

    @Override
    public Optional<E> findOne(ID id)
    {
        logger.traceEntry("Finding entity with ID: " + id);
        String sql = getFindOneSQL();
        try (PreparedStatement statement = getConnection().prepareStatement(sql))
        {
            setFindOneDeleteParameters(statement, id);
            try (ResultSet rs = statement.executeQuery())
            {
                if (rs.next())
                {
                    E entity = extractEntity(rs);
                    logger.traceExit("Entity found: " + entity);
                    return Optional.of(entity);
                }
            }
        }
        catch (SQLException e)
        {
            logger.error("Error finding entity with ID: " + id, e);
            throw new PersistenceException("Error finding entity", e);
        }
        logger.traceExit("No entity found with ID: " + id);
        return Optional.empty();
    }

    @Override
    public Optional<E> save(E entity)
    {
        logger.traceEntry("Saving entity: " + entity);
        String sql = getSaveSQL();
        try (PreparedStatement statement = getConnection().prepareStatement(sql, java.sql.Statement.RETURN_GENERATED_KEYS))
        {
            setSaveParameters(statement, entity);
            statement.executeUpdate();
            try (ResultSet generatedKeys = statement.getGeneratedKeys())
            {
                if (generatedKeys.next())
                {
                    int newId = generatedKeys.getInt(1);
                    entity.setId((ID) Integer.valueOf(newId));
                }
            }
            logger.traceExit("Entity saved successfully: " + entity);
            return Optional.of(entity);
        }
        catch (SQLException e)
        {
            logger.error("Error saving entity: " + entity, e);
            throw new PersistenceException("Error saving entity", e);
        }
    }

    @Override
    public Optional<E> delete(ID id)
    {
        logger.traceEntry("Deleting entity with ID: " + id);
        Optional<E> entityToDelete = findOne(id);
        if (entityToDelete.isPresent())
        {
            String sql = getDeleteSQL();
            try (PreparedStatement statement = getConnection().prepareStatement(sql))
            {
                setFindOneDeleteParameters(statement, id);
                statement.executeUpdate();
                logger.traceExit("Entity deleted successfully: " + entityToDelete.get());
                return entityToDelete;
            }
            catch (SQLException e)
            {
                logger.error("Error deleting entity with ID: " + id, e);
                throw new PersistenceException("Error deleting entity", e);
            }
        }
        logger.traceExit("No entity found to delete with ID: " + id);
        return Optional.empty();
    }

    @Override
    public Optional<E> update(E entity)
    {
        logger.traceEntry("Updating entity: " + entity);
        String sql = getUpdateSQL();
        try (PreparedStatement statement = getConnection().prepareStatement(sql))
        {
            setUpdateParameters(statement, entity);
            int rowsAffected = statement.executeUpdate();
            if (rowsAffected > 0)
            {
                logger.traceExit("Entity updated successfully: " + entity);
                return Optional.of(entity);
            }
        }
        catch (SQLException e)
        {
            logger.error("Error updating entity: " + entity, e);
            throw new PersistenceException("Error updating entity", e);
        }
        logger.traceExit("No entity updated for: " + entity);
        return Optional.empty();
    }

    @Override
    public int size()
    {
        logger.traceEntry("Getting size of entities");
        String sql = getSizeSQL();
        try (PreparedStatement statement = getConnection().prepareStatement(sql);
             ResultSet rs = statement.executeQuery())
        {
            if (rs.next())
            {
                int count = rs.getInt(1);
                logger.traceExit("Size of entities: " + count);
                return count;
            }
        }
        catch (SQLException e)
        {
            logger.error("Error counting entities", e);
            throw new PersistenceException("Error counting entities", e);
        }
        logger.traceExit("Size of entities: 0");
        return 0;
    }

    public void openConnection()
    {
        try
        {
            connection = DriverManager.getConnection(url);
            logger.info("Database connection established successfully");
        }
        catch (SQLException e)
        {
            logger.error("Error establishing database connection", e);
            throw new PersistenceException("Error establishing database connection", e);
        }
    }

    public void closeConnection()
    {
        try
        {
            if (connection != null && !connection.isClosed())
            {
                connection.close();
                logger.info("Database connection closed successfully");
            }
        }
        catch (SQLException e)
        {
            logger.error("Error closing database connection", e);
        }
    }

    protected abstract String getFindAllSQL();
    protected abstract String getFindOneSQL();
    protected abstract String getSaveSQL();
    protected abstract String getDeleteSQL();
    protected abstract String getUpdateSQL();
    protected abstract String getSizeSQL();

    protected abstract E extractEntity(ResultSet rs) throws SQLException;
    protected abstract void setSaveParameters(PreparedStatement statement, E entity) throws SQLException;
    protected abstract void setFindOneDeleteParameters(PreparedStatement statement, ID id) throws SQLException;
    protected abstract void setUpdateParameters(PreparedStatement statement, E entity) throws SQLException;
}

File: Anul 2/Medii de Programare si Proiectare/mpp-proiect-csharp-Mothyque/ProiectMPP/ProiectMPP.Server/Repository/DBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using log4net;
using ProiectMPP.Common.Domain;

namespace ProiectMPP.Server.Repository
{
    public abstract class DBRepository<TID, TE> : IRepository<TID, TE> where TE : Entity<TID>
    {
        protected readonly string connectionString;
        protected static readonly ILog log = LogManager.GetLogger(typeof(DBRepository<TID, TE>));
        protected IDbConnection connection;
        protected IDbTransaction transaction;

        public DBRepository(string connectionString)
        {
            this.connectionString = connectionString;
        }

        protected IDbConnection GetConnection()
        {
            return SharedConnectionManager.GetConnection(connectionString);
        }

        public void BeginTransaction()
        {
            SharedConnectionManager.BeginTransaction(connectionString);
            transaction = SharedConnectionManager.GetCurrentTransaction();
        }

        public void CommitTransaction()
        {
            try
            {
                SharedConnectionManager.CommitTransaction();
                transaction = null;
            }
            catch (Exception ex)
            {
                RollbackTransaction();
                throw;
            }
        }

        public void RollbackTransaction()
        {
            try
            {
                SharedConnectionManager.RollbackTransaction();
                transaction = null;
            }
            catch (Exception ex)
            {
            }
        }

        public void CloseConnection()
        {
            if (connection != null && connection.State != ConnectionState.Closed)
            {
                connection.Close();
                connection.Dispose();
                connection = null;
            }
        }

        public TE FindOne(TID id)
        {
            string sql = GetFindOneSQL();
            try
            {
                using (var command = GetConnection().CreateCommand())
                {
                    command.CommandText = sql;
                    if (transaction != null) command.Transaction = transaction;
                    SetDeleteParameters(command, id);
                    using (var reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            return ExtractEntity(reader);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                throw;
            }
            return null;
        }

        public IEnumerable<TE> FindAll()
        {
            List<TE> entities = new List<TE>();
            string sql = GetFindAllSQL();
            try
            {
                using (var command = GetConnection().CreateCommand())
                {
                    command.CommandText = sql;
                    if (transaction != null) command.Transaction = transaction;
                    using (var reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            entities.Add(ExtractEntity(reader));
                        }
                    }
                }
                return entities;
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        public TE Save(TE entity)
        {
            string sql = GetSaveSQL();
            try
            {
                using (var command = GetConnection().CreateCommand())
                {
                    command.CommandText = sql;
                    if (transaction != null) command.Transaction = transaction;
                    SetSaveParameters(command, entity);
                    command.ExecuteNonQuery();
                    return entity;
                }
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        public TE Delete(TID id)
        {
            TE entityToDelete = FindOne(id);
            if (entityToDelete != null)
            {
                string sql = GetDeleteSQL();
                try
                {
                    using (var command = GetConnection().CreateCommand())
                    {
                        command.CommandText = sql;
                        if (transaction != null) command.Transaction = transaction;
                        SetDeleteParameters(command, id);
                        command.ExecuteNonQuery();
                        return entityToDelete;
                    }
                }
                catch (Exception ex)
                {
                    throw;
                }
            }
            return null;
        }

        public TE Update(TE entity)
        {
            string sql = GetUpdateSQL();
            try
            {
                using (var command = GetConnection().CreateCommand())
                {
                    command.CommandText = sql;
                    if (transaction != null) command.Transaction = transaction;
                    SetUpdateParameters(command, entity);
                    int rowsAffected = command.ExecuteNonQuery();
                    return rowsAffected > 0 ? entity : null;
                }
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        public int Size()
        {
            string sql = GetSizeSQL();
            try
            {
                using (var command = GetConnection().CreateCommand())
                {
                    command.CommandText = sql;
                    if (transaction != null) command.Transaction = transaction;
                    return Convert.ToInt32(command.ExecuteScalar());
                }
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        protected abstract string GetFindAllSQL();
        protected abstract string GetFindOneSQL();
        protected abstract string GetSaveSQL();
        protected abstract string GetDeleteSQL();
        protected abstract string GetUpdateSQL();
        protected abstract string GetSizeSQL();
        protected abstract TE ExtractEntity(IDataReader reader);
        protected abstract void SetSaveParameters(IDbCommand command, TE entity);
        protected abstract void SetDeleteParameters(IDbCommand command, TID id);
        protected abstract void SetUpdateParameters(IDbCommand command, TE entity);
    }
}